### PR TITLE
add a new mux type, clean up docs DEVX-3930

### PR DESCRIPTION
**Pull request by brandon-ja**
This changes Puffin to use an internal mux for multiplexing commands to the correct handler functions. Previously, this was handled by the `funcMap` but that solution was significantly less flexible. Making this change required a lot of reworking of the code internals and some interesting additions as well.
1) I added a `bins` member to the FuncExec. This is because `Patterns` need not explicitly match a specific command (which was previously true with the `funcMap`). This new `bins` dir serves as the pseudo `$PATH` for the `Exec`
2) I added '*' matching in several places including the `Pattern` implementations as well as the `bins` member.
3) I added an internal `pat` type which is used to replicate the functionality of the `funcMap` and prevents a circular import of the new `pkg/pat` package.
4) I added a `pkg/pat` package to implement the new `Pattern` interface for external users of puffin.
5) I added the `Pattern` interface to server as a matcher for the mux. I heavily referenced the [goji mux](https://pkg.go.dev/goji.io) for this.
6) I had to update a lot of tests to use the new mux rather than the func map.
7) I updated the README with some simpler code that's now possible due to the new additions. (more docs still need to be written but this is sufficient for now I think)

### CODE REVIEW

File: func.go [hunk 1]
```diff
@@ -130,6 +161,23 @@ func WithEnv(envs map[string]string) FuncExecOption {
...
+	return func(fExec *FuncExec) {
+		if fExec.mux == nil {
+			fExec.mux = NewMux()
+		}
+		fExec.mux.HandleFunc(pat("*"), fn)
+	}
+}
+
+// WithMux sets a command Mux to the funcExec
+func WithMux(mux *Mux) FuncExecOption {
```
Comment (brandon-ja) [inline func.go]: Why have these as functional options? Are these really optional? Now that we have a unified `mux` type that we should require the FuncExec always gets a mux. We can expose simple muxes like `Success` or `Fail` as well as supporting the function map or a more robust mux.

File: func.go [hunk 2]
```diff
@@ -89,37 +116,41 @@ func (e *FuncExec) CommandContext(ctx context.Context, name string, arg ...strin
...
-// FuncExecOption can be used to configure the FuncExec struct
-type FuncExecOption func(*FuncExec)
-
-// WithFuncMap sets the func map used by all the commands created by this Exec
-func WithFuncMap(funcs map[string]CmdFunc) FuncExecOption {
+// WithBins adds the specified bins to the pseudo $PATH used by this Exec
+func WithBins(bins ...string) FuncExecOption {
 	return func(fExec *FuncExec) {
-		fExec.funcMap = funcs
+		fExec.bins = append(fExec.bins, bins...)
```
Comment (brandon-ja) [inline func.go:127]: Should this still be a functional option or should it be a variadic to the NewFuncExec function?

**Review by carsonoid — DISMISSED** [review]

File: func.go [hunk 3]
```diff
@@ -14,41 +15,69 @@ import (
...
 func (e *FuncExec) LookPath(file string) (string, error) {
-	_, found := e.findFunc(file)
-	if found != "" {
-		return found, nil
-	}
+	hasWildcard := slices.Contains(e.bins, "*")
 
 	if strings.Contains(file, "/") {
-		return found, &exec.Error{Name: file, Err: &os.PathError{Op: "stat", Path: file, Err: syscall.ENOENT}}
+		for _, p := range e.bins {
```
Comment (brandon-ja) [inline func.go]: use slices library

File: func.go [hunk 4]
```diff
@@ -14,41 +15,69 @@ import (
...
+			if file == p {
+				return p, nil
+			}
+		}
+
+		// if nothing else matches but bins has * matcher still return a match
+		if hasWildcard {
```
Comment (brandon-ja) [inline func.go:62]: simplify?

File: func.go [hunk 5]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"golang.org/x/exp/slices"
```
Comment (oo-weave) [inline func.go]: resort import

**Review by oo-weave — APPROVED** [review]

**Pull request merged**